chore: recursion howto explainers (#3792)
noir-lang/noir: docs/docs/explainers/explainer-recursion.md

## Some architecture

As with everything in computer science, there's no one-size-fits all. But there are some patterns that could help understanding and implementing them. To give three examples:

### Adding some logic to a proof verification

This would be an approach for something like our guessing game, where proofs are sent back and forth and are verified by each opponent. This circuit would be divided in two sections:

- A `recursive verification` section, which would be just the call to `std::verify_proof`, and that would be skipped on the first move (since there's no proof to verify)
- A `guessing` section, which is basically the logic part where the actual guessing happens

In such a situation, and assuming Alice is first, she would skip the first part and try to guess Bob's number. Bob would then verify her proof on the first section of his run, and try to guess Alice's number on the second part, and so on.

> One could be confused about what happens with the recursive aggregation object, as it seems like it has no relationship with anything other than the "recursive part" of the circuit (it's simply an input and an output to `std::verify_proof`). It actually doesn't, even if it may seem lightly unintuitive. Again, it would help to imagine it as a baton in a draft race, which says nothing about the race itself but simply testifies that it was correctly passed around.

Reviewer: I am not sure what you are trying to say here.

Author: Hard to describe, it stems from my own confusion as discussed internally with Tom and Maxim.

Maybe it's worth removing unless we have clear indication of people also being confused like I was

### Aggregating proofs

In some one-way interaction situations, recursiveness would allow for aggregation of simple proofs that don't need to be immediately verified on-chain or elsewhere.

To give a practical example, a barman wouldn't need to verify a "proof-of-age" on-chain every time he serves alcohol to a customer. Instead, the architecture would comprise two circuits:

- A `main`, non-recursive circuit with some logic
- A `recursive` circuit meant to verify two proofs in one proof

> The customer's proofs would be intermediate, and made on their phones, and the barman could just verify them locally. He would then aggregate them into a final proof sent on-chain (or elsewhere) at the end of the day.

Reviewer: I think this example is kind of confusing without more context. Why would the barman need to send proofs on-chain (or anywhere)?

Author: Well he would be able to prove that he only served people above drinking age. I guess the motivation isn't too important here 😂

keywords:
  [
    "Recursive Proofs",
    "Zero-Knowledge Programming",
    "Noir",
    "EVM Blockchain",
    "Smart Contracts",
    "Recursion in Noir",
    "Alice and Bob Guessing Game",
    "Recursive Merkle Tree",
    "Reusable Components",
    "Optimizing Computational Resources",
    "Improving Efficiency",
    "Verification Key",
    "Aggregation Objects",
    "Recursive zkSNARK schemes",
    "PLONK",
    "Proving and Verification Keys"
  ]
sidebar_position: 1
---

In programming, we tend to think of recursion as something calling itself. A classic example would be the calculation of the factorial of a number:

```js
function factorial(n) {
    if (n === 0 || n === 1) {
        return 1;
    } else {
        return n * factorial(n - 1);
    }
}
```

In this case, while `n` is not `1`, this function will keep calling itself until it hits the base case, bubbling up the result on the call stack:

```md
        Is `n` 1?  <---------
           /\               /
          /  \         n = n -1 
         /    \           /
       Yes     No --------
```

In Zero-Knowledge, recursion has some similarities.

> It is not a Noir function calling itself, but a proof being used as an input to another circuit. In short, you verify one proof *inside* another proof, returning the proof that both proofs were valid.

Reviewer: ```suggestion
It is not a Noir function calling itself, but a proof being used as an input to another circuit. In short, you verify one proof *inside* another proof, returning the proof that both proofs are valid.
```

> This can mean that, given enough computational resources, you can prove the correctness of any arbitrary number of proofs before, for example, settling them on-chain. This could be useful to design state channels (for which a common example would be [Bitcoin's Lightning Network](https://en.wikipedia.org/wiki/Lightning_Network)), to save on gas costs, or simply to make business logic less dependent on a consensus mechanism.

Reviewer 2: ```suggestion
This can mean that, given enough computational resources, you can prove the correctness of any arbitrary number of proofs in a single proof. This could be useful to design state channels (for which a common example would be [Bitcoin's Lightning Network](https://en.wikipedia.org/wiki/Lightning_Network)), to save on gas costs by settling one proof on-chain, or simply to make business logic less dependent on a consensus mechanism.
```
What do you think about rewriting like this?

Author: Yes, definitely 💯

> This can mean that, given enough computational resources, you can prove the correctness of any arbitrary number of proofs in a single proof. This could be useful to design state channels (for which a common example would be [Bitcoin's Lightning Network](https://en.wikipedia.org/wiki/Lightning_Network)), to save on gas costs by settling one proof on-chain, or simply to make business logic less dependent on a consensus mechanism.

Reviewer: ```suggestion
This means that, given enough computational resources, you can prove the correctness of any arbitrary number of proofs in a single proof. This could be useful to design state channels (for which a common example would be [Bitcoin's Lightning Network](https://en.wikipedia.org/wiki/Lightning_Network)), to save on gas costs by settling one proof on-chain, or simply to make business logic less dependent on a consensus mechanism.
```

## Examples

Let us look at some of these examples

### Alice and Bob - Guessing game

> Alice and Bob are friends, and they like guessing games. They want to play a guessing game online, but for that, they need an trusted third-party that knows both of their secrets and finishes the game once someone wins.

Reviewer 2: ```suggestion
Alice and Bob are friends, and they like guessing games. They want to play a guessing game online, but for that, they need a trusted third-party that knows both of their secrets and finishes the game once someone wins.
```

So, they use zero-knowledge proofs. Alice tries to guess Bob's number, and Bob will generate a ZK proof stating whether she succeeded or failed.

> This ZK proof can go on a smart contract, revealing the winner and even giving prizes. However, this means every turn needs to be verified on-chain. This incurs in some cost and waiting times that may simply turn the game too expensive or time-consuming to be worth it.

Reviewer 2: ```suggestion
This ZK proof can go on a smart contract, revealing the winner and even giving prizes. However, this means every turn needs to be verified on-chain. This incurs some cost and waiting time that may simply make the game too expensive or time-consuming to be worth it.
```

### Daniel - Reusable components

Daniel has a big circuit and a big headache. A part of his circuit is a setup phase that finishes with some assertions that need to be made. But that section alone takes 40% (80sec) of the 200sec proving time, and is largely independent of the rest of the circuit.

He could find it more efficient to generate a proof for that setup phase separately, and verifying it his actual business logic section of the circuit. For example, at the time of writing, on a Macbook M2, a recursive proof would take about 40 seconds using all the 8 cores in parallel.

This would make Daniel save 40 seconds by leveraging recursive proofs.

## What params do I need

As you can see in the [recursion reference](noir/standard_library/recursion.md), a simple recursive proof requires:

- The proof to verify
- The Verification Key of the circuit that generated the proof
- A hash of this verification key, as it's needed for some backends
- The public inputs for the proof
- The input aggregation object

It also returns the `output aggregation object`. These aggregation objects can be confusing at times, so let's dive in a little bit.

> ### Aggregation objects

Reviewer 2: This is good to explain right now, but we are going to be changing this very soon (https://github.com/AztecProtocol/aztec-packages/pull/3528) where the aggregation objects are hidden from the user. Do you think it is still worth it to include this section?

Author: Well tbh this page was on my to-do list for weeks now, so I reckon I'm a bit late to the party now 😁  but I'd still merge it like this and change it later, as this PR would still be useful to lay down what we have and will make it easier to change later on

We can imagine the `aggregation object` as the baton in a [relay race](https://en.wikipedia.org/wiki/Relay_race). The first runner doesn't have to receive the baton from anyone else, as he/she already starts with it. But when his/her turn is over, the next runner needs to receive it, run a bit more, and pass it along. Even though every runner could theoretically verify the baton mid-run (why not? 🏃🔍), only at the end of the race does the referee verify that the whole race is valid.

## How many circuits do I need

This is a common question, as there's not a one-size-fits-all solution. Here are three out of the many possible configurations:

### Two, but watch out

A common pattern, that requires some resources, is to have everyone generate both a main proof and a recursive proof. This involves two circuits:

- A `main`, non-recursive circuit with some logic
- A `recursive` circuit meant to verify two proofs in one proof

In this case, if Alice and Bob are playing a chess game and Alice is playing whites, she would immediately start by sending Bob a recursive proof, by verifying her `main` proof twice to start with.

> From then on, Bob would generate his `main` proof and verify both Alice's recursive proof, and his own `main` proof. Everyone generates two proofs, which could be intensive.

Reviewer 3: I don't follow this verification strategy. Why would you need to generate two proofs?

I would handle this scenario with a single proof which validates the previous step and then applies my own move to the board. Each player would just have to generate their own proof using the proof of the other player and then pass it back to their opponent.

Author: Yes that's "the light approach" described below. The problem here is the first proof, as there's no previous proof to verify. So either you use three circuits like described below, or you verify the first proof twice

### The light approach

> In some one-way interaction situations, one could shave off one of the recursive proofs. This would fit, for example, a situation where the first prover is using a phone, which is currently unfit to generate recursive proofs in an acceptable time frame. With this configuration, only the first, lighter proof, is generated on the phone.

Reviewer 3: I'm not seeing the need for the `first proof` circuit here. Can you explain?

Author: Yeha this is the same answer as above :) I'll refactor it to be a bit clearer!

- A `main`, non-recursive circuit with some logic
- A `first proof` recursive circuit meant to verify only the first `main` circuit
- A `recursive` circuit meant to verify two proofs in one proof

> To give a practical example, a barman wouldn't need to verify a "proof-of-age" on-chain every time he serves alcohol to a costumer. These proofs would be made on the costumer's phones, and the barman would just verify them locally, aggregating them into a final proof sent on-chain at the end of the day. However, the very first costumer needs its own circuit, as there's only one proof to verify.

Reviewer 3: ```suggestion
To give a practical example, a barman wouldn't need to verify a "proof-of-age" on-chain every time he serves alcohol to a customer. These proofs would be made on the customer's phones, and the barman would just verify them locally, aggregating them into a final proof sent on-chain at the end of the day. However, the very first customer needs its own circuit, as there's only one proof to verify.
```

This barman has something against people who make costumes. :laughing:

Author: Sorry, my english sucks 😔

So, the tallier puts all the votes in a merkle tree, and everyone can also prove the verification of two proofs within one proof, as such:

```md
                    abcd
           __________|______________
          |                         |
         ab                         cd 
     _____|_____              ______|______
    |           |            |             |              
  alice        bob        charlie        daniel 
```

Doing this recursively allows us to arrive on a final proof `abcd` which if true, verifies the correctness of all the votes.

### Daniel - Reusable components

Daniel has a big circuit and a big headache. A part of his circuit is a setup phase that finishes with some assertions that need to be made. But that section alone takes 40% (80sec) of the 200sec proving time, and is largely independent of the rest of the circuit.

He could find it more efficient to generate a proof for that setup phase separately, and verifying it his actual business logic section of the circuit. For example, at the time of writing, on a Macbook M2, a recursive proof would take about 40 seconds using all the 8 cores in parallel.

> This would make Daniel save 40 seconds by leveraging recursive proofs.

Reviewer 3: I don't think that we need to give concrete numbers here as this dates the explainer. Instead we should say that this allows this proof to be constructed in parallel with the proof with the rest of his business logic as this is what actually allows a speedup.

- A `recursive verification` section, which would be just the call to `std::verify_proof`, and that would be skipped on the first move (since there's no proof to verify)
- A `guessing` section, which is basically the logic part where the actual guessing happens

In such a situation, and assuming Alice is first, she would skip the first part and try to guess Bob's number. Bob would then verify her proof on the first section of his run, and try to guess Alice's number on the second part, and so on.

One could be confused about what happens with the recursive aggregation object, as it seems like it has no relationship with anything other than the "recursive part" of the circuit (it's simply an input and an output to `std::verify_proof`). It actually doesn't, even if it may seem lightly unintuitive. Again, it would help to imagine it as a baton in a draft race, which says nothing about the race itself but simply testifies that it was correctly passed around.

### The light approach

In some one-way interaction situations, one could actually use two circuits, in order to make proving faster for one of the parties.

> This would fit, for example, a situation where the first prover is using a phone. Phones are currently unfit to generate recursive proofs in an acceptable time frame, given the gate size of the circuit. With this configuration, only the first, lighter proof, is generated on the phone.

Reviewer 3: This architecture isn't used to get around phone processing speeds.

One would, then, use two circuits:

- A `main`, non-recursive circuit with some logic
- A `recursive` circuit meant to verify two proofs in one proof

> To give a practical example, a barman wouldn't need to verify a "proof-of-age" on-chain every time he serves alcohol to a customer. These proofs would be made on the customer's phones, and the barman would just verify them locally, aggregating them into a final proof sent on-chain at the end of the day. However, the very first customer needs its own circuit, as there's only one proof to verify.

Reviewer 3: The way that this is written (e.g. "the light approach") implies that these two architectures are two different ways to do the same thing but they're not.

The first is incremental computation where you have a single state which is progressed through each verification whereas the second is just simple aggregation where each proof is independent of each other.

Reviewer 3: I bring this up as we have different sections for different usecases above (Alice + BoB vs Charlie.) but then we merge them all together here.

Author: Makes sense, I'll present them as three examples instead
